### PR TITLE
 Fix glTF morph target animation stutter from shader recompiles

### LINKED ISSUE #18594: glTF morph-target animations stutter when an influence animates through zero
## Problem

When a glTF/glb file is loaded with an animation that drives morph target weights, the mesh visibly stutters every time one of the animated influences passes through exactly `0`. On real content this shows up as a brief one-frame "snap" — for example, a character's mouth snapping fully closed and back open partway through the clip.

## What is happening

The morph vertex shader is specialized for a specific number of influencers via the `NUM_MORPH_INFLUENCERS` define, which is derived from `morphTargetManager.numMaxInfluencers || morphTargetManager.numInfluencers`.

The glTF loader creates the `MorphTargetManager` with the defaults `numMaxInfluencers = 0` and `optimizeInfluencers = true`, so the define follows the *live* number of currently non-zero influencers. As an animated influence drops to `0`, that target is removed from the active set, the active influencer count changes, the define changes, and the material is recompiled. While the new shader compiles, the submesh reports `isReady() === false` and is skipped for that frame, so the mesh disappears for a frame — that dropped frame is the visible stutter. Because the influence keeps crossing zero as the animation plays, this recompile/skip happens repeatedly.

This is loader-specific: the glTF loader never pins `numMaxInfluencers`, so any glTF/glb with morph-weight animation that crosses zero is affected. The FBX loader already sets `optimizeInfluencers = false` and `numMaxInfluencers = numTargets`, which is why FBX content does not exhibit the stutter.

## What is expected

The morph animation should play smoothly. The morph shader should be compiled once for the mesh and should not be recompiled mid-animation simply because an animated influence value passes through zero.

### PR DESCRIPTION
**Pull request by PatrickRyanMS**
The glTF loader created the MorphTargetManager with the defaults (numMaxInfluencers = 0, optimizeInfluencers = true), so NUM_MORPH_INFLUENCERS tracked the live count of non-zero influencers. When an animated morph influence passed through zero, that target dropped out of the active set, the influencer count changed, the shader define changed, and the material recompiled. While the new shader compiled, the submesh reported isReady() === false and was skipped for a frame, producing a visible one-frame "snap" (e.g. a character's mouth snapping shut and back open) on morph-animated meshes.

 Pin the influencer count when loading morph targets by setting optimizeInfluencers = false and numMaxInfluencers = numTargets, matching the FBX loader and the existing core helpers, so the morph shader is compiled once. Both are required so the active influencer set - and the bound influences/attributes - always matches NUM_MORPH_INFLUENCERS, including the vertex-attribute fallback path, which has no per-frame influencer-count guard.

 The behavior is gated behind a new glTF loader option, useMaxMorphTargetInfluencers (default true), so apps with very large blendshape rigs can opt out and keep the previous per-frame optimization.

 Fixes #18594.

 Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

**Review by sebavan — APPROVED** [review]

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18596/merge

Test environment:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18596/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18596/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18596/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18596/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18596/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**13** passed, **0** leaked out of **13** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (13)</summary>

| Scenario | Package |
|----------|---------|
| Core Feature Stack | @babylonjs/core |
| Core Rendering Materials Shadows Stack | @babylonjs/core |
| Core Textures Render Targets PostProcess Stack | @babylonjs/core |
| GUI Fullscreen UI Controls | @babylonjs/gui |
| GUI Mesh ADT Controls | @babylonjs/gui |
| Loaders Boombox Import | @babylonjs/loaders |
| Loaders OBJ Direct Load | @babylonjs/loaders |
| Loaders STL Direct Load | @babylonjs/loaders |
| Materials Library Stack | @babylonjs/materials |
| Serializers glTF Export | @babylonjs/serializers |
| Serializers GLB Export | @babylonjs/serializers |
| PostProcesses Digital Rain Stack | @babylonjs/post-processes |
| Procedural Textures Stack | @babylonjs/procedural-textures |

</details>

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

File: packages/dev/loaders/test/integration/babylon.sceneLoader.test.ts [hunk 1]
```diff
@@ -398,6 +398,46 @@ test.describe("Babylon Scene Loader", function () {
...
+                });
+                return BABYLON.SceneLoader.ImportMeshAsync(null, "https://assets.babylonjs.com/meshes/Alien/", "Alien.gltf", window.scene).then(() => {
+                    BABYLON.SceneLoader.OnPluginActivatedObservable.remove(observer);
+                    const manager = (window.scene!.getMeshByName("AlienHead") as any).morphTargetManager;
+                    return {
+                        numMaxInfluencers: manager.numMaxInfluencers,
+                        optimizeInfluencers: manager.optimizeInfluencers,
+                    };
+                });
+            });
```
[the remark this answers is not in the document]
Comment (sebavan) [reply packages/dev/loaders/test/integration/babylon.sceneLoader.test.ts:435]: @PatrickRyanMS you could use addOnce instead

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**13** passed, **0** leaked out of **13** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (13)</summary>

| Scenario | Package |
|----------|---------|
| Core Feature Stack | @babylonjs/core |
| Core Rendering Materials Shadows Stack | @babylonjs/core |
| Core Textures Render Targets PostProcess Stack | @babylonjs/core |
| GUI Fullscreen UI Controls | @babylonjs/gui |
| GUI Mesh ADT Controls | @babylonjs/gui |
| Loaders Boombox Import | @babylonjs/loaders |
| Loaders OBJ Direct Load | @babylonjs/loaders |
| Loaders STL Direct Load | @babylonjs/loaders |
| Materials Library Stack | @babylonjs/materials |
| Serializers glTF Export | @babylonjs/serializers |
| Serializers GLB Export | @babylonjs/serializers |
| PostProcesses Digital Rain Stack | @babylonjs/post-processes |
| Procedural Textures Stack | @babylonjs/procedural-textures |

</details>

File: packages/dev/loaders/src/glTF/2.0/glTFLoader.ts [hunk 2]
```diff
@@ -1342,6 +1342,16 @@ export class GLTFLoader implements IGLTFLoader {
...
+
+            if (this._parent.useMaxMorphTargetInfluencers) {
+                // Compile the morph shader once for all targets so it is not recompiled (causing a one-frame
+                // visual glitch) when an animated influence passes through zero and the active influencer count
+                // changes. optimizeInfluencers must be disabled too so the active influencer set - and thus the
+                // bound influences/attributes - always matches NUM_MORPH_INFLUENCERS, including in the
+                // vertex-attribute fallback path (which has no per-frame influencer-count guard).
+                morphTargetManager.optimizeInfluencers = false;
+                morphTargetManager.numMaxInfluencers = morphTargetManager.numTargets;
+            }
```
[the remark this answers is not in the document]
Comment (PatrickRyanMS) [reply packages/dev/loaders/src/glTF/2.0/glTFLoader.ts:1358]: added a guard for numMaxInfluencers/optimizeInfluencers

File: packages/dev/loaders/test/integration/babylon.sceneLoader.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (PatrickRyanMS) [reply packages/dev/loaders/test/integration/babylon.sceneLoader.test.ts:435]: Added addOnce for the test as well as added a second test to validate the edge case of forcing attribute mode and exceeding influencer cap leaving influencers unpinned.

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18596/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**13** passed, **0** leaked out of **13** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (13)</summary>

| Scenario | Package |
|----------|---------|
| Core Feature Stack | @babylonjs/core |
| Core Rendering Materials Shadows Stack | @babylonjs/core |
| Core Textures Render Targets PostProcess Stack | @babylonjs/core |
| GUI Fullscreen UI Controls | @babylonjs/gui |
| GUI Mesh ADT Controls | @babylonjs/gui |
| Loaders Boombox Import | @babylonjs/loaders |
| Loaders OBJ Direct Load | @babylonjs/loaders |
| Loaders STL Direct Load | @babylonjs/loaders |
| Materials Library Stack | @babylonjs/materials |
| Serializers glTF Export | @babylonjs/serializers |
| Serializers GLB Export | @babylonjs/serializers |
| PostProcesses Digital Rain Stack | @babylonjs/post-processes |
| Procedural Textures Stack | @babylonjs/procedural-textures |

</details>
